include/openssl/evp.h

```diff
@@ -690,20 +690,49 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
 // containing the result. It returns one on success or zero on error.
 OPENSSL_EXPORT int EVP_PKEY_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **out_pkey);
 
-OPENSSL_EXPORT int EVP_PKEY_encapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
-OPENSSL_EXPORT int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx,
-                                        unsigned char *wrapped_key,
-                                        size_t *wrapped_key_len,
-                                        unsigned char *gen_key,
-                                        size_t *gen_key_len);
-OPENSSL_EXPORT int EVP_PKEY_decapsulate_init(EVP_PKEY_CTX *ctx,
-                                             const OSSL_PARAM params[]);
-OPENSSL_EXPORT int EVP_PKEY_decapsulate(EVP_PKEY_CTX *ctx,
-                                        unsigned char *unwrapped,
-                                        size_t *unwrapped_len,
-                                        const unsigned char *wrapped,
-                                        size_t wrapped_len);
+// EVP_PKEY_encapsulate is an operation defined for a KEM (Key Encapsulation
+// Mechanism). For the KEM specified in |ctx|, the function:
+//   1. determines the size of the shared secret and the ciphertext and
+//      writes them to |shared_secret_len| and |ciphertext_len|,
+//   2. generates a random value of size |shared_secret_len| and
+//      writes it to |shared_secret|,
+//   3. encapsulates the shared secret, producing the ciphertext, by using
+//      the public key in |ctx| and writes the ciphertext to |ciphertext|.
+//
+// If the given |ciphertext| is NULL it is assumed that the caller is doing
+// a size check: the function will write the size of the ciphertext and the
+// shared secret in |ciphertext_len| and |shared_secret_len| and return 1.
+//
+// NOTE: no allocation is done in the function, the caller is expected to
+// provide large enough |ciphertext| and |shared_secret| buffers.
```

Comment:
Do other EVP functions check/enforce this or just rely on customers doing the right thing and testing with valgrind/address sanitizer?

Do we have tests for bad input sizes? Do we have a SIM for fuzz tests?

Reply:
> Do other EVP functions check/enforce this or just rely on customers doing the right thing and testing with valgrind/address sanitizer?

I haven't gone through all the EVP function, but the ones I did go through check if the `*_len` value is large enough. Not sure what you mean by "enforcing"?

> Do we have tests for bad input sizes? Do we have a SIM for fuzz tests?

I added a test for too small input sizes in `p_kyber_test.cc`. The SIM for fuzz tests: CryptoAlg-1295.